### DISCUSSION TITLE
About to create RSA key and certificate in optee using PKCS#11 tool in optee side host

### DISCUSSION THREAD

**Discussion by pavanchate** [strongswan/strongswan#1660] [Q&A]
I am wondering to create the RSA key and  certificate in optee using pkcs11 tool .
I am able to establish the ipsec connection in between two hosts. But in one host I am using optee to load the private key and certificate from the token not from the specific path.

For that I have used the following commands to create the certificate in one directory :
pki --gen --type rsa --size 2048 --outform pem > strongswankey.pem

 pki --self --ca --lifetime 3652 --in strongswankey.pem --dn "C=CH, O=strongSwan, CN=strongSwan Root CA" --outform pem > strongswanCert.pem
 
 
pki --gen --type rsa --size 3072 --outform pem> sunkey.pem

pki --gen --type rsa --size 3072 --outform pem> moonkey.pem

pki --req --type priv --in sunkey.pem --dn "C=CH, O=strongswan, CN=sun.strongswan.org" --san sun.strongswan.org --outform pem > sunReq.pem

pki --issue --cacert strongswanCert.pem --cakey strongswankey.pem --type pkcs10 --in sunReq.pem --serial 01 --lifetime 1826 --outform pem > sunCert.pem

pki --issue --cacert strongswanCert.pem --cakey strongswankey.pem --type pkcs10 --in moonReq.pem --serial 01 --lifetime 1826 --outform pem > moonCert.pem

And I have used the following commands to created the RSA key in token : :

pkcs11-tool --module /usr/lib/libckteec.so.0 --init-token --label mytoken --so-pin 4321 --init-pin --pin 1234

pkcs11-tool --module /usr/lib/libckteec.so.0 --token-label mytoken --login --pin 1234 --keypairgen --label mykey --id 4114 --key-type rsa:2048

Now, I am wondering to craete the RSA certificate and that should be  load from that token only.

could you please help me further ?

**Comment by tobiasbrunner** [comment]
I guess you are referring to OP-TEE?

Anyway, instead of using `pki --gen` to generate a key, you directly use the key on the token with the `--req` command. Configure the module for the [_pkcs11_ plugin](https://docs.strongswan.org/docs/5.9/plugins/pkcs11.html) and then pass the key's ID via `--keyid` to the command. To use the private key on the token during authentication, configure a [`secrets.token*` section in swanctl.conf](https://docs.strongswan.org/docs/5.9/swanctl/swanctlConf.html#_secrets_tokensuffix).

**Comment by pavanchate** [comment]
Yes, I am referring to optee. And  I want to load the certificate and private key from the token. can you help me with the example of how to generate the token and put key and certificate in token with the commands in optee side?

**Reply by tobiasbrunner** [reply]
Just use the information I provided above. And you might be able to get some information from [the TPM 2.0 tutorial](https://docs.strongswan.org/docs/5.9/tpm/tpm2.html), which stores keys and certificates on the TPM that's accessed like a PKCS#11 token (obviously not directly comparable as different tools and plugins are used and key IDs might have a different format etc., but you see the `pki` commands that may be used).

**Reply by pavanchate** [reply]
Do I need any kind of hardware to store the key and certificate or directly we can put the RSA key and certificate in the token and load from that itself ? for strong swan connection in between two hosts.

**Reply by tobiasbrunner** [reply]
You can load certificates directly from the token (the _pkcs11_ plugin actually does so automatically, by default). They can be associated with a config via identity or explicitly via `local.cert*` section (you can see that in [the TPM 2 config example](https://docs.strongswan.org/docs/5.9/tpm/tpm2Ike.html)). Keys are never actually loaded, they are just accessed via PKCS#11 to create signatures, so you don't need any extra storage for those either.

**Reply by pavanchate** [reply]
Ok thanks for you information. I will go through with the information you have provided. 
If get some kind of need then again contact you here only. Thank you very much Tobias Brunner.

**Comment by pavanchate** [comment]
Hi,
I am creating the following certificate in linux standalone optee system by using following commands::
#pki --gen --type rsa --size 2048 --outform pem > strongswankey.pem
#pki --self --ca --lifetime 3652 --in strongswankey.pem --dn "C=CH, O=strongSwan, CN=strongSwan Root CA" --outform pem > strongswanCert.pem
#pki --gen --type rsa --size 3072 --outform pem> sunkey.pem
#pki --gen --type rsa --size 3072 --outform pem> moonkey.pem
#pki --req --type priv --in sunkey.pem --dn "C=CH, O=strongswan, CN=sun.strongswan.org" --san sun.strongswan.org --outform pem > sunReq.pem
#pki --issue --cacert strongswanCert.pem --cakey strongswankey.pem --type pkcs10 --in sunReq.pem --serial 01 --lifetime 1826 --outform pem > sunCert.pem
# pki --issue --cacert strongswanCert.pem --cakey strongswankey.pem --type pkcs10 --in moonreq.pem --serial 01 --lifetime 1826 --outform pem > moonCert.pem

Step 2: 
Creating csr.cnf file in same directory by adding  below parameter in csr.cnf::

[req]
distinguished_name = req_distinguished_name
req_extensions = v3_req
prompt = no

[req_distinguished_name]
C = CH
O = strongswan
CN = moon.strongswan.org

[v3_req]
subjectAltName = DNS:moon.strongswan.org


STEP 3: 
I am creating the token for RSA key and RSA certificate by using following commands::

#pkcs11-tool --module /usr/lib/libckteec.so.0 --init-token --label mytoken --so-pin 4321 --init-pin --pin 1234
#pkcs11-tool --module /usr/lib/libckteec.so.0 --token-label mytoken --login --pin 1234 --keypairgen --label mykey --id 4114 --key-type rsa:2048

STEP 4:
Using openssl :

#engine dynamic -pre SO_PATH:/usr/lib/engines-1.1/pkcs11.so -pre ID:pkcs11 -pre LIST_ADD:1 -pre LOAD -pre MODULE_PATH:/usr/lib/libckteec.so.0
# req -new -config csr.cnf -engine pkcs11 -key pkcs11:object=mykey;type=private?pin-value=1234 -keyform engine -outform pem -out moonreq.pem 
# openssl req -new -keyform engine -engine pkcs11 -key "pkcs11:token=mytoken;object=mykey;pin-value=1234" -subj "/CN=moonCert" -out moonCert.pem

I am trying to craete  the  above commands to create the certificate in the token, which should  be loaded from optee and the IPSec connection should be  up in between two host.

I am very new in this I am not able to craete the RSA certificate and  key in the optee using token and not able  to load the connection from optee.

Could you please help me  w.r.t above information with appropriate guidance to make it done?

**Reply by tobiasbrunner** [reply]
> #pki --gen --type rsa --size 3072 --outform pem> sunkey.pem

Again, if you want keys on the token, don't use `pki --gen` but instead use whatever tool to create keys on the token and then use `pki --req` to create a signature request based on that key (via `--keyid` option).

> Step 2:
> Creating csr.cnf file in same directory by adding below parameter in csr.cnf::

Why use `openssl` if you already started using `pki`? I can't help you with `openssl`.

**Comment by pavanchate** [comment]
I have followed the below steps in optee host to create the RSA key and certificate in optee from token::
And also I have given the output of each command and at the end I am getting the error while creating the creating the certificate, could ypu please analyze the below information and provide me the exact solution? 


# export SLOT=0
# export LABEL="my_token"
# export MODULE="/usr/lib/libckteec.so.0.1.0"
# pkcs11-tool --show-info --module /usr/lib/libckteec.so.0  2>&1 | tee -a /tmp/q
emuoptee.log
Using slot 0 with a present token (0x0)
Cryptoki version 2.40
Manufacturer     Linaro
Library          OP-TEE PKCS11 Cryptoki library (ver 0.1)
# pkcs11-tool --list-slots --module /usr/lib/libckteec.so.0 2>&1 | tee -a /tmp/q
emuoptee.log
Available slots:
Slot 0 (0x0): OP-TEE PKCS11 TA - TEE UUID 94e9ab89-4c43-56ea-8b35-45dc07226830
  token state:   uninitialized
Slot 1 (0x1): OP-TEE PKCS11 TA - TEE UUID 94e9ab89-4c43-56ea-8b35-45dc07226830
  token state:   uninitialized
Slot 2 (0x2): OP-TEE PKCS11 TA - TEE UUID 94e9ab89-4c43-56ea-8b35-45dc07226830
  token state:   uninitialized
# export CKTEEC_LOGIN_TYPE="public"
# pkcs11-tool --init-token --label my_token --so-pin "1234" --slot=0 --module /u
sr/lib/libckteec.so.0 2>&1 | tee -a /tmp/qemuoptee.log
Token successfully initialized
# pkcs11-tool --list-slots --module /usr/lib/libckteec.so.0 2>&1 | tee -a /tmp/q
emuoptee.log
Available slots:
Slot 0 (0x0): OP-TEE PKCS11 TA - TEE UUID 94e9ab89-4c43-56ea-8b35-45dc07226830
  token label        : my_token
  token manufacturer : Linaro
  token model        : OP-TEE TA
  token flags        : login required, rng, token initialized
  hardware version   : 0.0
  firmware version   : 0.1
  serial num         : 0000000000000000
  pin min/max        : 4/128
Slot 1 (0x1): OP-TEE PKCS11 TA - TEE UUID 94e9ab89-4c43-56ea-8b35-45dc07226830
  token state:   uninitialized
Slot 2 (0x2): OP-TEE PKCS11 TA - TEE UUID 94e9ab89-4c43-56ea-8b35-45dc07226830
  token state:   uninitialized
# pkcs11-tool --login --login-type so --so-pin "any_so_pin" --test --token-label
 my_token --module /usr/lib/libckteec.so.0 2>&1 | tee -a /tmp/qemuoptee.log
error: PKCS11 function C_Login failed: rv = CKR_PIN_INCORRECT (0xa0)
Aborting.
# 
# pkcs11-tool --login --login-type so --so-pin "4114" --test --token-label my_to
ken --module /usr/lib/libckteec.so.0 2>&1 | tee -a /tmp/qemuoptee.log
error: PKCS11 function C_Login failed: rv = CKR_PIN_INCORRECT (0xa0)
Aborting.
# pkcs11-tool --login --login-type so --so-pin "1234" --test --token-label my_to
ken --module /usr/lib/libckteec.so.0 2>&1 | tee -a /tmp/qemuoptee.log
C_SeedRandom() and C_GenerateRandom():
  seems to be OK
Digests:
  all 4 digest functions seem to work
  MD5: OK
  SHA-1: OK
Signature: not a R/W session, skipping signature tests
Verify: not a R/W session, skipping verify tests
Decryption: not a R/W session, skipping decryption tests
No errors

# pkcs11-tool --login --so-pin "1234" --init-pin --pin "4321" --token-label my_t
oken --module /usr/lib/libckteec.so.0 2>&1 | tee -a /tmp/qemuoptee.log
User PIN successfully initialized
# pkcs11-tool --list-slots --module /usr/lib/libckteec.so.0 2>&1 | tee -a /tmp/q
emuoptee.log
Available slots:
Slot 0 (0x0): OP-TEE PKCS11 TA - TEE UUID 94e9ab89-4c43-56ea-8b35-45dc07226830
  token label        : my_token
  token manufacturer : Linaro
  token model        : OP-TEE TA
  token flags        : login required, rng, token initialized, PIN initialized
  hardware version   : 0.0
  firmware version   : 0.1
  serial num         : 0000000000000000
  pin min/max        : 4/128
Slot 1 (0x1): OP-TEE PKCS11 TA - TEE UUID 94e9ab89-4c43-56ea-8b35-45dc07226830
  token state:   uninitialized
Slot 2 (0x2): OP-TEE PKCS11 TA - TEE UUID 94e9ab89-4c43-56ea-8b35-45dc07226830
  token state:   uninitialized

# pkcs11-tool --login --login-type user --pin "4321" --test --token-label my_tok
en --module /usr/lib/libckteec.so.0 2>&1 | tee -a /tmp/qemuoptee.log
C_SeedRandom() and C_GenerateRandom():
  seems to be OK
Digests:
  all 4 digest functions seem to work
  MD5: OK
  SHA-1: OK
Signatures: not implemented
Verify (currently only for RSA)
  No private key found for testing
Decryption (currently only for RSA)
No errors
# pkcs11-tool --login --login-type user --pin "4321" --keypairgen --key-type RSA
:2048 --id 03 --label rsa_key --token-label my_token --module /usr/lib/libckteec
.so.0 2>&1 | tee -a /tmp/qemuoptee.log
Key pair generated:
Private Key Object; RSA 
  label:      rsa_key
  ID:         03
  Usage:      decrypt, sign
  Access:     sensitive, always sensitive, never extractable, local
Public Key Object; RSA 2048 bits
  label:      rsa_key
  ID:         03
  Usage:      encrypt, verify
  Access:     local

# openssl req -engine pkcs11 -new -key "pkcs11:token=my_token;object=rsa_key;typ
e=private;pin-value=4321" -keyform engine -out "rsa_cert.pem" -text -x509 -days 
358000 -subj "/CN=localhost" 2>&1 | tee -a /tmp/qemuoptee.log
engine "pkcs11" set.
Unable to load module (null)
PKCS11_get_private_key returned NULL
cannot load Private Key from engine
1995558944:error:81065401:libp11:pkcs11_CTX_load:Unable to load PKCS#11 module:p11_load.c:77:
1995558944:error:8006B067:pkcs11 engine:func(107):invalid parameter:eng_back.c:603:
1995558944:error:26096080:engine routines:ENGINE_load_private_key:failed loading private key:crypto/engine/eng_pkey.c:77:
unable to load Private Key

**Reply by tobiasbrunner** [reply]
Again, I can't help you with problems with `openssl`.
